internal/civisibility/utils/net/client.go

```diff
@@ -0,0 +1,228 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand/v2"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/constants"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/utils"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+const (
+	DefaultMaxRetries int           = 5
+	DefaultBackoff    time.Duration = 150 * time.Millisecond
+)
+
+type (
+	Client interface {
+		GetSettings() (*SettingsResponseData, error)
+		GetEarlyFlakeDetectionData() (*EfdResponseData, error)
+		GetCommits(localCommits []string) ([]string, error)
+		SendPackFiles(packFiles []string) (bytes int64, err error)
+	}
+
+	client struct {
+		id                 string
+		agentless          bool
+		baseUrl            string
+		environment        string
+		serviceName        string
+		workingDirectory   string
+		repositoryUrl      string
+		commitSha          string
+		branchName         string
+		testConfigurations testConfigurations
+		headers            map[string]string
+		handler            *RequestHandler
+	}
+
+	testConfigurations struct {
+		OsPlatform          string            `json:"os.platform,omitempty"`
+		OsVersion           string            `json:"os.version,omitempty"`
+		OsArchitecture      string            `json:"os.architecture,omitempty"`
+		RuntimeName         string            `json:"runtime.name,omitempty"`
+		RuntimeArchitecture string            `json:"runtime.architecture,omitempty"`
+		RuntimeVersion      string            `json:"runtime.version,omitempty"`
+		Custom              map[string]string `json:"custom,omitempty"`
+	}
+)
+
+func NewClient() Client {
+	ciTags := utils.GetCITags()
+
+	// get the environment
+	environment := os.Getenv("DD_ENV")
+	if environment == "" {
+		environment = "none"
+	}
+
+	// get the service name
+	serviceName := os.Getenv("DD_SERVICE")
+	if serviceName == "" {
+		if repoURL, ok := ciTags[constants.GitRepositoryURL]; ok {
+			// regex to sanitize the repository url to be used as a service name
+			repoRegex := regexp.MustCompile(`(?m)/([a-zA-Z0-9\\\-_.]*)$`)
```

Comment:
Why `\` is included in the valid characters?

Also, wouldn't it make more sense to accept one or more of anything except `/` in the repo name?

Reply:
I see, but if I understand correctly, in the linked code the backslash is escaping the `-` from being interpreted as a regex character range. Here the backslashes have been duplicated. (If I understand correctly, `@"..."` in C# has similar behavior to `` `...` `` in Go.)

Reply:
changed in [f0c2d4c](https://github.com/DataDog/dd-trace-go/pull/2869/commits/f0c2d4c412c042cb29af4fbe5f9824420f452d42)

Reply:
Looks like this commit got lost :sweat_smile:

Reply:
oh the rebase :( pushed here: [d2119cb](https://github.com/DataDog/dd-trace-go/pull/2869/commits/d2119cbc46ee3bbde3134ad06c70c4b41c04353b)